src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -539,56 +539,73 @@ internal override Compilation WithEventQueue(AsyncQueue<CompilationEvent> eventQ
-            if (lastStatement == null || !lastStatement.Statement.IsKind(SyntaxKind.ExpressionStatement))
+            if (HasTopLevelReturnStatements(root))
             {
-                return GetSpecialType(SpecialType.System_Void);
+                return true;
             }
 
-            var expressionStatement = (ExpressionStatementSyntax)lastStatement.Statement;
-            if (!expressionStatement.SemicolonToken.IsMissing)
+            // Is there a trailing expression?
+            var lastGlobalStatement = (GlobalStatementSyntax)root.Members.LastOrDefault(m => m.IsKind(SyntaxKind.GlobalStatement));
+            if (lastGlobalStatement != null)
             {
-                return GetSpecialType(SpecialType.System_Void);
+                var statement = lastGlobalStatement.Statement;
+                if (statement.IsKind(SyntaxKind.ExpressionStatement))
+                {
+                    var expressionStatement = (ExpressionStatementSyntax)statement;
+                    if (expressionStatement.SemicolonToken.IsMissing)
+                    {
+                        var model = GetSemanticModel(tree);
+                        var expression = expressionStatement.Expression;
+                        var info = model.GetTypeInfo(expression);
+                        return ((TypeSymbol)info.ConvertedType)?.SpecialType != SpecialType.System_Void;
+                    }
+                }
             }
 
-            var model = GetSemanticModel(tree);
-            hasValue = true;
-            var expression = expressionStatement.Expression;
-            var info = model.GetTypeInfo(expression);
-            return (TypeSymbol)info.ConvertedType;
+            return false;
+        }
+
+        private static bool HasTopLevelReturnStatements(SyntaxNode node)
+        {
+            Debug.Assert(node.SyntaxTree.Options.Kind == SourceCodeKind.Script);
+            return node.DescendantNodes(ShouldCheckDescendantsForReturns).Any(n => n.IsKind(SyntaxKind.ReturnStatement));
+        }
+
+        private static bool ShouldCheckDescendantsForReturns(SyntaxNode node)
+        {
+            switch (node.Kind())
+            {
+                // These are not top level statement contexts.
+                case SyntaxKind.ClassDeclaration:
+                case SyntaxKind.StructDeclaration:
+                case SyntaxKind.DelegateDeclaration:
+                case SyntaxKind.EnumDeclaration:
+                case SyntaxKind.InterfaceDeclaration:
+                case SyntaxKind.MethodDeclaration:
+                case SyntaxKind.AnonymousMethodExpression:
+                case SyntaxKind.ParenthesizedLambdaExpression:
+                case SyntaxKind.SimpleLambdaExpression:
+                    return false;
+                default:
+                    return true;
+            }
```

Comment:
I think that would still descend into lambdas (which we don't want).  Per Tomas's suggestion, I think I'm going to try ```!(node is DeclarationSyntax || node is Expression)```.